[CPU][ArmSME] Enable transposes for f32 and f64

c-rhodes opened the pull request:
This patch enables transposes for ArmSME for f32 and f64 types now that they
may get eliminated by folding into defining `vector.transfer_read` since [1].
This is done by extending `setTransposeLikeOpRootConfig` which currently only
supports x86 (no changes there). The transpose is represented by a
`linalg.generic`, since `linalg.transpose` gets converted by
`GeneralizeLinalgNamedOps`.

[1] https://github.com/llvm/llvm-project/pull/92562

ci-extra: build_test_all_arm64

c-rhodes committed:
[CPU][ArmSME] Enable transposes for f32 and f64

Since [1] transposes may get eliminated by folding into defining
vector.transfer_read.

[1] https://github.com/llvm/llvm-project/pull/92562

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

MaheshRavishankar reviewed (commented):
Would be good to wait for either @hanhanW or @dcaballe review.

MacDue commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1907,43 +1907,58 @@ setTransposeLikeOpRootConfig(mlir::FunctionOpInterface entryPointFn,
+        distConfig.minTileSizes.begin(), distConfig.minTileSizes.end(),
+        [](int64_t tileSize) { return tileSize > 1; }, targetVectorSize);
+  } else if (targetIsAArch64SME) {
+    auto elementType = nonWideningLinalgElementType(genericOp);
```
I wonder if this should be more restrictive? This seems to set 2-D scalable tile sizes for any transpose-like operation, but not all of them can/do lower to ArmSME, for example, I tired out:
```
  %empty = tensor.empty(): tensor<2x16x16xf32>
  %rhs = linalg.transpose ins(%transposed_b : tensor<2x16x16xf32>)
      outs(%empty : tensor<2x16x16xf32>) permutation = [0, 2, 1]
```
...it set 2-D scalable flags but ended up unvectorized (and used linalg-to-loops in the end).

c-rhodes replied:
absolutely, I misinterpreted `x86TransposeLoweringPrecondition` to be restricted to 2 dims. I've kept that as is and created a separate one for SME to be more restrictive, test also added.

c-rhodes committed:
add more restrictions

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

c-rhodes committed:
return failure for unsupported target

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1980,46 +2005,67 @@ setTransposeLikeOpRootConfig(mlir::FunctionOpInterface entryPointFn,
+  bool targetIsAArch64SME = isAArch64(targetAttr) &&
+                            hasSMEFeature(targetAttr) &&
+                            clEnableScalableVectorization;
+  if (hasAVX2Feature(targetAttr)) {
```
The big if-else makes readability really bad. Can we create methods to get `minTileSizes`? So it can be like `std::optional<SmallVector<int64_t>> getX86TransposeVectorSizes()` and `std::optional<SmallVector<int64_t>> getAArch64TransposeVectorSizes()`. What do you think?

c-rhodes committed:
add transpose vector size methods for targets

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

MacDue commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -334,6 +334,31 @@ static bool x86TransposeLoweringPrecondition(linalg::GenericOp genericOp) {
+  // Check all iterators are parallel.
+  if (genericOp.getNumParallelLoops() != genericOp.getNumLoops())
+    return false;
+
```
I think this also needs to check the body of the generic is simply a `scf.yield %in`. 
I tried: 
```
linalg.generic {indexing_maps = [affine_map<(d0, d1) -> (d1, d0)>, affine_map<(d0, d1) -> (d0, d1)>], iterator_types = ["parallel", "parallel"]} ins(%2 : tensor<32x32xf32>) outs(%3 : tensor<32x32xf32>) {
^bb0(%in: f32, %out: f32):
  %mul = arith.mulf %in, %in : f32
  linalg.yield %mul : f32
} -> tensor<32x32xf32>
```

And that's still a 'transpose'  and it adds the scalable flags and you get: 
```
%11 = arith.mulf %9, %9 : vector<[4]x[4]xf32>
%12 = vector.transpose %11, [1, 0] : vector<[4]x[4]xf32> to vector<[4]x[4]xf32>
%13 = vector.transfer_write %12, %extracted_slice_0[%c0, %c0], %10 {in_bounds = [true, true]} : vector<[4]x[4]xf32>, tensor<?x?xf32>
```

MacDue replied:
(`opKnownToSupport2DScalableVectorizationWithArmSME` would also need to check this)

c-rhodes replied:
good spot! I've updated to check for this, but had to move this function to utils to also call it from `opKnownToSupport2DScalableVectorizationWithArmSME`.

banach-space reviewed (commented):
Thanks Cullen!

Sorry for the delay, I've kept skipping this one without realising that this is _the patch_ :) I blame the title:
> [CPU][ArmSME] Enable transposes for f32 and f64

That's not what's happening here, is it? IIUC, this patch adds logic to identify transpose-like `linalg.generic` Ops and then, for ArmSME, adds "scalable" flags to the tiling config.  For X86 there should be no changes, right? 

That's my understanding based on the code - is this correct? If yes, could you update the summary?

c-rhodes committed:
Check generic body has single yield

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
+  if (elementType->isF64()) {
+    sizes.append({2, 2});
+    scalableSizeFlags.append({true, true});
+  }
```
I think that you also need to make sure that for element types other than `f32` and `f64`, the pre-condition fails. In the spirit of failing early.

MacDue replied:
_technically_ this fails no later than if it were in the pre-condition (but avoids duplicating it there, so I think this is fine).

c-rhodes replied:
As Ben points out, doing this would require duplicating the logic which isn't ideal, particularly now that the precondition isn't specific to SME, so going to leave this as is.

banach-space reviewed (approved):
LGTM, modulo handling other data types, thanks!

MacDue reviewed (approved):
LGTM :+1: Thanks for the changes!

MacDue commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,44 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
   return hasFeature(targetAttr, "+i8mm");
 }
 
+bool transposeLoweringPreconditionAArch64SME(linalg::GenericOp genericOp) {
```
I think calling this `transposeLoweringPreconditionAArch64SME` makes this look more specific to ArmSME than it really is. This is just something like `isLinalgGeneric2DTransposition`.

hanhanW replied:
+1, I don't see any ArmSME specifics in the method.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,44 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
+    return false;
+  auto yieldOp = dyn_cast<linalg::YieldOp>(body.getTerminator());
+  if (!yieldOp)
+    return false;
```
nit: I think we can use `cast` because it should already be verified by LinalgOp.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,44 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
+
+  // Check op has single input and output.
+  if ((genericOp.getNumDpsInputs() != 1) || (genericOp.getNumDpsInits() != 1))
+    return false;
```
nit: we don't need `()` for the two conditions.

```suggestion
  if (genericOp.getNumDpsInputs() != 1 || genericOp.getNumDpsInits() != 1)
    return false;
```

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,44 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
 }
 
+bool transposeLoweringPreconditionAArch64SME(linalg::GenericOp genericOp) {
+  // Check op has 2 dimensions.
+  if (genericOp.getNumLoops() != 2)
+    return false;
+
+  // Check op has single input and output.
```
nit: remove the comments, because the code already describes what they are doing. Same comment for line 70.

https://google.github.io/styleguide/cppguide#Implementation_Comment_Donts

c-rhodes replied:
personally find they're much quicker to parse, so I've them.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,44 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
+      return false;
+    if (yieldArg.getArgNumber() != 0)
+      return false;
+  }
```
Given that we already check that the number of inits is one, so we don't need the for loop here. (I believe that people won't expect that there are more than one value, so it's fine.)

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
       [](int64_t tileSize) { return tileSize > 1; }, targetVectorSize);
 
+  sizes = minTileSizes;
+  scalableSizeFlags = SmallVector<bool>(sizes.size(), false);
```
This kind of logics keep showing up in targets that don't care about scalable vectors. I did not get time to revamp the idea/design of large refactoring with @banach-space that I mentioned in https://github.com/iree-org/iree/pull/16435/files#r1513504805 (my bad). We really need a plan and take action items to address the issue. What do you think about it? (I'm just flagging the concern, but not saying that it is a blocker.)

c-rhodes replied:
strongly agree targets that don't support scalability shouldn't have to care about it, sloppiness on my part. I've updated the target agnostic `getTransposeVectorSizes` to assume empty scalable flags indicates the target doesn't care, so target-specific hooks like this don't have to deal with it.

banach-space replied:
> We really need a plan and take action items to address the issue. What do you think about it?

@hanhanW 100% agree, sorry for not being more pro-active about this :( I was waiting for a bit more feedback, but I think that the overall suggestion was to try a bit more substantial refactor?

hanhanW replied:
(no action item required, just open the conversation for visibility.)

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
+  if (failed(elementType))
+    return;
+
+  if (elementType->isF32()) {
```
nit: use `.isF32`. same for `.isF64`.

c-rhodes replied:
> nit: use `.isF32`. same for `.isF64`.

doesn't compile

hanhanW replied:
oh I see, it is because that `nonWideningLinalgElementType` returns `FailureOr<Type>` but not `Type`. I misread it, sorry for the bad suggestion.

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
-      !x86TransposeLoweringPrecondition(genericOp)) {
-    return failure();
-  }
+/// Utility to compute the transpose vector sizes for X86.
```
Need more details about the documentation. Without looking into implementation, it is not clear that the vectors are empty when it fails. Also, how about returning a boolean value which indicates if it succeeds or not?

c-rhodes replied:
I quite like the empty state of `sizes` indicating failure so I've kept that, but updated the docs to make it clear 👍

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
+                               targetMLTransInfo, tileSizes, scalableFlags);
+  else if (isAArch64(targetAttr) && hasSMEFeature(targetAttr) &&
+           clEnableScalableVectorization)
+    getTransposeAArch64SMEVectorSizes(genericOp, tileSizes, scalableFlags);
```
few nits:

1. It is better to use `isX86` for the first condition.
2. style nit: wrap `braces` for the if-else statements. This is the few difference between IREE and MLIR in terms of coding style.
3. Can we move the `isSMEFeature` and `clEnableScalableVectorization` logics to `getTransposeAArch64SMEVectorSizes`, and rename it to `getTransposeAArch64VectorSizes`? Today, most of the methods are separated based on the `isX86`, `isRISCV`, and `isAArch64`. IMO, keeping it this way is better for layering. I'm afraid that this if-statements becomes a monster and adds complexities.

c-rhodes replied:
much nicer 👍

c-rhodes committed:
address comments

- In `getTransposeVectorSizes` assume target doesn't care about
scalability if scalable flags are empty. Target specific functions like
`getTransposeX86VectorSizes` can then ignore scalable flags.
- Document `sizes` may be empty on return from
`getTransposeX86VectorSizes` indicating failure.
- add braces.
- Predicate `getTransposeX86VectorSizes` on `isX86` and move
`hasAVX2Feature` check to `getTransposeX86VectorSizes`.
- s/getTransposeAArch64SMEVectorSizes/getTransposeAArch64VectorSizes.
- Predicate `getTransposeAArch64VectorSizes` on `isAArch64` and move SME
checks to `getTransposeAArch64VectorSizes`.
- s/transposeLoweringPreconditionAArch64SME/isLinalgGeneric2DTranspose.
- Replace loop checking yield op returns block arg corresponding to
input with check on single val.

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

MacDue commented on compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPU2DScalableTo1DScalable.cpp:
```diff
@@ -87,6 +88,9 @@ class LLVMCPU2DScalableTo1DScalablePass
 static bool opKnownToSupport2DScalableVectorizationWithArmSME(Operation *op) {
+  if (auto genericOp = dyn_cast<linalg::GenericOp>(op))
+    return LinalgOpInfo(genericOp).isTranspose() &&
+           isLinalgGeneric2DTranspose(genericOp);
```
`LinalgOpInfo` is redundant right? 
```suggestion
    return isLinalgGeneric2DTranspose(genericOp);
```

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/Utils.cpp:
```diff
@@ -58,4 +58,39 @@ bool hasI8mmFeature(IREE::HAL::ExecutableTargetAttr targetAttr) {
+
+  auto yieldOp = cast<linalg::YieldOp>(body.getTerminator());
+
+  // The yield op should return the block argument corresponds to the input.
```
```suggestion
  // The yield op should return the block argument corresponding to the input.
```

banach-space commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,127 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
+  sizes = minTileSizes;
+}
+
+/// Utility to return the transpose vector `sizes` for AArch64. Empty `sizes` on
+/// return indicates failure.
```
This _will_ return empty sizes for SVE - what's the plan for SVE here? Perhaps just add a TODO?

c-rhodes replied:
I've added a note that only SME is supported. The focus of this patch is on SME, this could be extended to support SVE.

c-rhodes committed:
address comments

remove redundant LinalgOpInfo isTranspose check.

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

c-rhodes committed:
address comments

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

hanhanW commented on compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:
```diff
@@ -1970,57 +1970,128 @@ setTransformStrategyRootConfig(mlir::FunctionOpInterface entryPointFn,
+/// return indicates failure.
+static void getTransposeX86VectorSizes(
+    linalg::GenericOp genericOp, IREE::HAL::ExecutableTargetAttr targetAttr,
+    SmallVectorImpl<int64_t> &minTileSizes, SmallVectorImpl<int64_t> &sizes) {
```
I think the type of minTileSizes can be `ArrayRef<int64_t>`. We don't expect that the it would be modified. At the last part, you can do something like:

```
sizes.assign(minTileSizes);
std::replace_if(sizes.begin(), sizes...
```

hanhanW reviewed (approved):
Thanks, just one nit! LGTM!

c-rhodes committed:
address comments

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>

c-rhodes committed:
run clang-format

Signed-off-by: Cullen Rhodes <cullen.rhodes@arm.com>
